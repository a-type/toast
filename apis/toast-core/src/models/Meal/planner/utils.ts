import { ScheduleMeal } from 'models/Schedule/Schedule';
import { Meal, Schedule } from 'models';

export const MEALS_PER_DAY = 3;
export const DAYS_PER_WEEK = 7;

export const mealsFromGroceryDay = (schedule: Schedule): ScheduleMeal[] => {
  return [
    ...schedule.meals.slice(schedule.groceryDay * MEALS_PER_DAY),
    ...schedule.meals.slice(0, schedule.groceryDay * MEALS_PER_DAY),
  ];
};

const findCookedMeal = meal =>
  meal.mealIndex !== 0 &&
  ['LONG', 'MEDIUM', 'SHORT'].includes(meal.availability);

export const initializeWeek = (
  schedule: Schedule,
  startDateIndex: number,
  endDateIndex: number,
  findFirstMeal: (meal: ScheduleMeal) => boolean = findCookedMeal,
): Meal[] => {
  // skip breakfast for main planning
  // find first cook day after grocery day to use to initialize the week
  const firstCookMeal = mealsFromGroceryDay(schedule).find(findFirstMeal);

  if (!firstCookMeal) {
    throw new Error(
      "Sorry, we couldn't figure out where to start this plan based on your schedule.",
    );
  }

  const meals = new Array(1 + endDateIndex - startDateIndex)
    .fill(null)
    .map((_, dayIndex) =>
      new Array(MEALS_PER_DAY).fill(null).map((__, mealIndex) => {
        return Meal.createEmpty(startDateIndex + dayIndex, mealIndex);
      }),
    )
    .reduce((allMeals, dayMeals) => allMeals.concat(dayMeals), []);

  return meals;
};
