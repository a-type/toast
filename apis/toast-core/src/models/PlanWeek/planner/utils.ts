import Schedule, { ScheduleMeal } from 'models/Schedule';
import { PlanWeekData } from '../types';

const MEALS_PER_DAY = 3;
const DAYS_PER_WEEK = 7;

export const mealsFromGroceryDay = (schedule: Schedule): ScheduleMeal[] => {
  return [
    ...schedule.meals.slice(schedule.groceryDay * MEALS_PER_DAY),
    ...schedule.meals.slice(0, schedule.groceryDay * MEALS_PER_DAY),
  ];
};

const findCookedMeal = meal =>
  meal.mealIndex !== 0 &&
  ['LONG', 'MEDIUM', 'SHORT'].includes(meal.availability);

export const initializeWeek = (
  schedule: Schedule,
  weekIndex: number,
  findFirstMeal: (meal: ScheduleMeal) => boolean = findCookedMeal,
) => {
  // skip breakfast for main planning
  // find first cook day after grocery day to use to initialize the week
  const firstCookMeal = mealsFromGroceryDay(schedule).find(findFirstMeal);

  if (!firstCookMeal) {
    throw new Error(
      "Sorry, we couldn't figure out where to start this plan based on your schedule.",
    );
  }

  const week: PlanWeekData = {
    id: `${schedule.id}_w_${weekIndex}`,
    weekIndex,
    mealsPerDay: MEALS_PER_DAY,
    startDay: firstCookMeal.dayIndex,
    startMeal: firstCookMeal.mealIndex,
    meals: new Array(MEALS_PER_DAY * DAYS_PER_WEEK)
      .fill(null)
      .map((_, idx) => ({
        id: `${schedule.id}_w_${weekIndex}_m_${idx}`,
        weekIndex,
        // offsets meal index according to the meal the week starts on
        mealIndex: (firstCookMeal.mealIndex + idx) % MEALS_PER_DAY,
        dayIndex: Math.floor(
          ((firstCookMeal.dayIndex * MEALS_PER_DAY +
            firstCookMeal.mealIndex +
            idx) /
            MEALS_PER_DAY) %
            DAYS_PER_WEEK,
        ),
        actions: [],
      })),
  };

  return week;
};
