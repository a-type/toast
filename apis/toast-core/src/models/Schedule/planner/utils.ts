import { ScheduleMeal } from 'models/Schedule/Schedule';
import { Meal, Schedule } from 'models';

export const MEALS_PER_DAY = 3;
export const DAYS_PER_WEEK = 7;

export const initializeWeek = (): Meal[] => {
  const meals = new Array(DAYS_PER_WEEK)
    .fill(null)
    .map((_, dayIndex) =>
      new Array(MEALS_PER_DAY).fill(null).map((__, mealIndex) => {
        return Meal.createEmpty(dayIndex, mealIndex);
      }),
    )
    .reduce((allMeals, dayMeals) => allMeals.concat(dayMeals), []);

  return meals;
};
