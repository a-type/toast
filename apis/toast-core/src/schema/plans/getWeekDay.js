import { addDays, addWeeks } from 'date-fns';

export default ({ weekIndex, startDay, dayOffset }) => {
  const date = new Date(startDay.valueOf());
  return addDays(addWeeks(date, weekIndex), dayOffset);
};
