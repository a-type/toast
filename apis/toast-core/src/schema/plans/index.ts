import { gql } from 'apollo-server-express';
import { Context } from 'context';
import { path, mergeDeepRight } from 'ramda';
import { id } from 'tools';

import * as meals from './meals';
import * as schedules from './schedules';

export const typeDefs = () => [
  gql`
    type Plan {
      id: ID!
    }

    extend type Query {
      plan: Plan
    }
  `,
  meals.typeDefs,
  schedules.typeDefs,
];

export const resolvers = [
  {
    Query: {
      plan: async (_parent, args, ctx: Context) => {
        const group = await ctx.graph.groups.getMine();
        let planId = path<string>(['planId'], group);

        if (!planId) {
          planId = id('plan');
          await ctx.graph.groups.mergeMine({ planId });
        }

        const plan = await ctx.firestore.plans.get(planId);

        return plan;
      },
    },
  },
  meals.resolvers,
  schedules.resolvers,
].reduce(mergeDeepRight, {});
