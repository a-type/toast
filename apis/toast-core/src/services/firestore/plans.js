const COLLECTION = 'plans';
const START_WEEK_DAY = new Date(2018, 10, 7);

const defaulted = plan => ({
  servingsPerMeal: 0,
  days: [],
  ...plan,
});

export default class Plans {
  static START_WEEK_DAY = START_WEEK_DAY;
  START_WEEK_DAY = START_WEEK_DAY;

  constructor(firestore) {
    this.firestore = firestore;
  }

  /**
   * For user's default plan, use user id
   */
  get = async planId => {
    const document = this.firestore.doc(`${COLLECTION}/${planId}`);
    const docRef = await document.get();
    if (!docRef.exists) {
      return null;
    }

    return defaulted(docRef.data());
  };

  merge = async (planId, planData) => {
    const document = this.firestore.doc(`${COLLECTION}/${planId}`);
    await document.set(planData, { merge: true });
    const docRef = await document.get();
    return defaulted(docRef.data());
  };

  set = async (planId, plan) => {
    const document = this.firestore.doc(`${COLLECTION}/${planId}`);
    await document.set(plan);
    return defaulted(plan);
  };

  /**
   * Weeks - copies of the base plan, applied per-week
   */
  getWeek = async (planId, week) => {
    const document = this.firestore.doc(`${COLLECTION}/${planId}/week_${week}`);
    const docRef = await document.get();

    if (!docRef.exists) {
      // copy the base plan instead
      return await this.get(planId);
    }

    return defaulted(docRef.data());
  };

  mergeWeek = async (planId, week, planData) => {
    const document = this.firestore.doc(`${COLLECTION}/${planId}/week_${week}`);
    await document.set(planData, { merge: true });
    const docRef = await document.get();
    return defaulted(docRef.data());
  };
}
