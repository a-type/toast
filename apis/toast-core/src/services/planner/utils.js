export const getNeededPrepServingsRanges = meals => {
  const neededPrepServingsRanges = [0];
  const incrementLast = () =>
    neededPrepServingsRanges[neededPrepServingsRanges.length - 1]++;
  meals.forEach(meal => {
    if (meal.availability === 'NONE') {
      incrementLast();
    } else if (meal.availability === 'LONG') {
      neededPrepServingsRanges.push(0);
    }
  });

  return neededPrepServingsRanges;
};

/**
 * "scroll" to the first instance of a particular set of availability in a
 * list of meals
 */
export const scroll = (list, scrollToAvailability = ['LONG']) => {
  const indexOfFirstLongDay = list.findIndex(meal =>
    scrollToAvailability.includes(meal.availability),
  );
  return [
    ...list.slice(indexOfFirstLongDay),
    ...list.slice(0, indexOfFirstLongDay),
  ];
};

export const getMealsByType = mealList => {
  ['long', 'medium', 'short', 'none'].reduce(
    (counts, type) => ({
      ...counts,
      [type]: mealList.filter(meal => meal.availability.toLowerCase() === type),
    }),
    {},
  );
};

export const toMealList = plan =>
  plan.days.reduce(
    (allMeals, day) => [...allMeals, day.breakfast, day.lunch, day.dinner],
    [],
  );

export const getPlanStrategy = plan => {
  let mealListWithoutBreakfast = toMealList(plan).filter(
    ({ meal }) => meal !== 'breakfast',
  );
  const mealsByType = getMealsByType(mealListWithoutBreakfast);

  if (mealsByType.long.length === 0 && mealsByType.none.length > 0) {
    return 'eatOutOnNones';
  }

  if (mealsByType.none.length === 0) {
    return 'basic';
  }

  if (
    mealsByType.long.length + mealsByType.medium.length >=
    mealsByType.none.length
  ) {
    return 'doublePortions';
  }

  if (
    mealsByType.long.length > 0 &&
    mealsByType.none.length / mealsByType.long.length <= 3
  ) {
    return 'prep';
  }

  if (
    mealsByType.long.length > 0 &&
    mealsByType.none.length / mealsByType.long.length <= 5
  ) {
    return 'bigPrep';
  }

  return 'bigPrepAndEatOut';
};

export const getMealById = (id, plan) => {
  const [day, meal] = id.split('.');
  const dayIdx = parseInt(day);
  return plan.days[dayIdx][meal];
};
