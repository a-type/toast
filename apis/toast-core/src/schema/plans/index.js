import { gql } from 'apollo-server-express';
import { id } from 'tools';
import { path, pathOr, assocPath, compose, mergeDeepLeft } from 'ramda';
import { UserInputError } from 'errors';
import getWeekIndex from './getWeekIndex';
import getWeekDay from './getWeekDay';
import Plan from 'services/planner/Plan';

const emptyPlan = {
  days: new Array(7).fill(null).map(() => ({
    meals: new Array(3).fill(null).map(() => ({
      availability: 'SKIP',
      actions: [],
    })),
  })),
};

export const typeDefs = gql`
  enum PrepAvailability {
    SKIP
    EAT_OUT
    NONE
    SHORT
    MEDIUM
    LONG
  }

  enum PlanActionType {
    EAT_OUT
    COOK
    EAT
    READY_MADE
    SKIP
  }

  enum PlanMealType {
    QUICK
    BIG
    FANCY
    NORMAL
  }

  enum PlanStrategy {
    BASIC
    PREP
    BIG_PREP
  }

  interface PlanAction {
    id: ID!
    type: PlanActionType!
    dayIndex: Int
    mealIndex: Int
  }

  type PlanActionEatOut implements PlanAction {
    id: ID!
    type: PlanActionType!
    dayIndex: Int
    mealIndex: Int
    note: String
  }

  type PlanActionCook implements PlanAction {
    id: ID!
    type: PlanActionType!
    dayIndex: Int
    mealIndex: Int
    servings: Int!
    mealType: PlanMealType!
    recipe: Recipe
  }

  type PlanActionEat implements PlanAction {
    id: ID!
    type: PlanActionType!
    dayIndex: Int
    mealIndex: Int
    leftovers: Boolean!
    cookAction: PlanActionCook
  }

  type PlanActionReadyMade implements PlanAction {
    id: ID!
    type: PlanActionType!
    dayIndex: Int
    mealIndex: Int
    note: String
  }

  type PlanActionSkip implements PlanAction {
    id: ID!
    type: PlanActionType!
    dayIndex: Int
    mealIndex: Int
  }

  type PlanMeal {
    id: ID!
    availability: PrepAvailability!
    actions: [PlanAction!]!
  }

  type PlanDay {
    id: ID!
    date: Date
    meals: [PlanMeal!]!
  }

  type Plan {
    id: ID!
    defaultServings: Int!
    days: [PlanDay!]!
    groceryDay: Int!
    warnings: [String!]!

    """
    Access weekly plans using the week index, which is based
    on the fixed plan chronology system. To calculate a week index
    for any particular day, use the root planWeekIndex query
    """
    week(weekIndex: Int!): Plan

    """
    A date representing the first day (Sunday) of this Plan week,
    if (and only if) this Plan is a weekly plan, and not the root "blueprint" plan
    """
    startDate: Date
  }

  input PlanSetDetailsInput {
    defaultServings: Int
    groceryDay: Int
  }

  input PlanSetMealDetailsInput {
    availability: PrepAvailability
  }

  extend type Mutation {
    setPlanDetails(details: PlanSetDetailsInput!): Plan!
      @hasScope(scope: "update:plan")
    setPlanMealDetails(
      dayIndex: Int!
      mealIndex: Int!
      details: PlanSetMealDetailsInput!
    ): Plan! @hasScope(scope: "update:plan")
    setPlanStrategy(strategy: PlanStrategy): Plan!

    setPlanActionRecipe(
      weekIndex: Int!
      dayIndex: Int!
      mealIndex: Int!
      actionId: ID!
      recipeId: ID!
    ): PlanAction!
  }

  extend type Group {
    plan: Plan @authenticated
  }

  extend type Query {
    """
    A shortcut function to determine the week index of any particular day
    within the planning system
    """
    planWeekIndex(year: Int!, month: Int!, date: Int!): Int!
    """
    The start date of the entire plan system. You can reference this
    to determine the chronology of plan weeks in conjunction with
    planWeekIndex and the weekIndex field on Plan itself
    """
    planStartWeekDate: Date!
  }
`;

export const resolvers = {
  Query: {
    planWeekIndex: (_parent, { year, month, date }, ctx) =>
      getWeekIndex({
        year,
        month,
        date,
        startDay: ctx.firestore.plans.START_WEEK_DAY,
      }),

    planStartWeekDate: (_parent, _args, ctx) =>
      ctx.firestore.plans.START_WEEK_DAY,
  },

  Mutation: {
    setPlanDetails: async (_parent, { details }, ctx) => {
      const group = await ctx.graph.groups.getMine();
      let planId = path(['planId'], group);
      // create plan if it doesn't exist
      if (!planId) {
        planId = id('plan');
        await ctx.graph.groups.mergeMine({ planId });
      }

      const defaultPlan = new Plan({ id: planId });
      const plan = (await ctx.firestore.plans.get(planId)) || defaultPlan;
      plan.defaultServings = details.defaultServings;
      plan.groceryDay = details.groceryDay;
      const processed = ctx.planner.run(plan);
      return ctx.firestore.plans.set(planId, processed);
    },

    setPlanMealDetails: async (
      _parent,
      { dayIndex, mealIndex, details },
      ctx,
    ) => {
      const group = await ctx.graph.groups.getMine();

      if (!group || !group.planId) {
        throw new UserInputError("You haven't created a plan yet");
      }

      const plan = await ctx.firestore.plans.get(group.planId);

      plan.setMealDetails(dayIndex, mealIndex, details);
      const processed = ctx.planner.run(plan);

      await ctx.firestore.plans.set(group.planId, plan);
      return plan;
    },

    setPlanStrategy: async (_parent, { strategy }, ctx) => {
      const group = await ctx.graph.groups.getMine();

      if (!group || !group.planId) {
        throw new UserInputError("You haven't created a plan yet");
      }

      const plan = await ctx.firestore.plans.get(group.planId);
      plan.strategy = strategy;
      const processed = ctx.planner.run(plan);
      await ctx.firestore.plans.set(group.planId, processed);
      return processed;
    },

    setPlanActionRecipe: async (
      _parent,
      { weekIndex, dayIndex, mealIndex, actionId, recipeId },
      ctx,
    ) => {
      const group = await ctx.graph.groups.getMine();

      if (!group || !group.planId) {
        throw new UserInputError("You haven't created a plan yet");
      }

      const plan = await ctx.firestore.plans.getWeek(group.planId, weekIndex);
      const action = plan.setActionRecipe(
        dayIndex,
        mealIndex,
        actionId,
        recipeId,
      );
      await ctx.firestore.plans.mergeWeek(group.planId, weekIndex, plan);
      return action;
    },
  },

  Group: {
    plan: async (parent, args, ctx) => {
      const { planId } = parent;
      const plan = await ctx.firestore.plans.get(planId);
      ctx.plan = plan;
      return plan;
    },
  },

  Plan: {
    week: async (parent, args, ctx) => {
      if (parent.weekIndex !== null) {
        throw new UserInputError(
          "You can't access the week field on a Plan which represents a week already",
        );
      }
      const week = await ctx.firestore.plans.getWeek(parent.id, args.weekIndex);
      ctx.week = week;
      return week;
    },

    startDate: (parent, _args, ctx) => {
      if (!parent.id.includes('week_')) {
        return null;
      }

      return getWeekDay({
        weekIndex: parent.weekIndex,
        startDay: ctx.firestore.plans.START_WEEK_DAY,
      });
    },
  },

  PlanAction: {
    __resolveType: obj => {
      switch (obj.type) {
        case 'COOK':
          return 'PlanActionCook';
        case 'EAT':
          return 'PlanActionEat';
        case 'EAT_OUT':
          return 'PlanActionEatOut';
        case 'READY_MADE':
          return 'PlanActionReadyMade';
        case 'SKIP':
          return 'PlanActionSkip';
      }
    },
  },

  PlanActionEat: {
    cookAction: async (parent, args, ctx) => {
      let plan = ctx.week || ctx.plan;
      if (!plan) {
        const group = await ctx.graph.groups.getMine();

        if (!group || !group.planId) {
          throw new UserInputError("You haven't created a plan yet");
        }

        plan = await ctx.firestore.plans.get(group.planId);
      }

      const { cookActionId } = parent;

      return plan.getAction(cookActionId);
    },
  },

  PlanActionCook: {
    recipe: async (parent, args, ctx) => {
      if (!parent.recipeId) {
        return null;
      }

      return ctx.graph.recipes.get(parent.recipeId);
    },
  },

  PlanDay: {
    date: (parent, args, ctx) => {
      if (ctx.week) {
        const weekIndex = ctx.week.weekIndex;
        const { dayIndex } = parent;

        const date = getWeekDay({
          weekIndex,
          startDay: ctx.firestore.plans.START_WEEK_DAY,
          dayOffset: dayIndex,
        });
        return date;
      }
      return null;
    },
  },
};
