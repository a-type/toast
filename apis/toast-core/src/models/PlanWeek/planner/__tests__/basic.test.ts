import plan from '../index';
import plans from './createPlans';
import { ignoreIds } from './utils';
const { variety } = plans();

describe('basic meal plan', () => {
  test('converts availability to meal types', () => {
    variety.strategy = 'BASIC';
    expect(ignoreIds(plan.run(variety, 0).toJSON())).toMatchSnapshot();
  });
});
