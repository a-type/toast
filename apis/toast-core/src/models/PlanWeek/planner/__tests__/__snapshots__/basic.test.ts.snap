// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic meal plan converts availability to meal types 1`] = `
Object {
  "meals": Array [
    Object {
      "actions": Array [
        Object {
          "dayIndex": 0,
          "mealIndex": 0,
          "type": "SKIP",
          "weekIndex": 0,
        },
      ],
      "dayIndex": 0,
      "mealIndex": 0,
      "weekIndex": 0,
    },
    Object {
      "actions": Array [
        Object {
          "dayIndex": 0,
          "mealIndex": 1,
          "mealType": "QUICK",
          "servings": 2,
          "type": "COOK",
          "weekIndex": 0,
        },
        Object {
          "dayIndex": 0,
          "leftovers": false,
          "mealIndex": 1,
          "type": "EAT",
          "weekIndex": 0,
        },
      ],
      "dayIndex": 0,
      "mealIndex": 1,
      "weekIndex": 0,
    },
    Object {
      "actions": Array [
        Object {
          "dayIndex": 0,
          "mealIndex": 2,
          "mealType": "FANCY",
          "servings": 2,
          "type": "COOK",
          "weekIndex": 0,
        },
        Object {
          "dayIndex": 0,
          "leftovers": false,
          "mealIndex": 2,
          "type": "EAT",
          "weekIndex": 0,
        },
      ],
      "dayIndex": 0,
      "mealIndex": 2,
      "weekIndex": 0,
    },
    Object {
      "actions": Array [
        Object {
          "dayIndex": 1,
          "mealIndex": 0,
          "type": "SKIP",
          "weekIndex": 0,
        },
      ],
      "dayIndex": 1,
      "mealIndex": 0,
      "weekIndex": 0,
    },
    Object {
      "actions": Array [
        Object {
          "dayIndex": 1,
          "mealIndex": 1,
          "type": "EAT_OUT",
          "weekIndex": 0,
        },
      ],
      "dayIndex": 1,
      "mealIndex": 1,
      "weekIndex": 0,
    },
    Object {
      "actions": Array [
        Object {
          "dayIndex": 1,
          "mealIndex": 2,
          "mealType": "NORMAL",
          "servings": 2,
          "type": "COOK",
          "weekIndex": 0,
        },
        Object {
          "dayIndex": 1,
          "leftovers": false,
          "mealIndex": 2,
          "type": "EAT",
          "weekIndex": 0,
        },
      ],
      "dayIndex": 1,
      "mealIndex": 2,
      "weekIndex": 0,
    },
    Object {
      "actions": Array [
        Object {
          "dayIndex": 2,
          "mealIndex": 0,
          "mealType": "QUICK",
          "servings": 2,
          "type": "COOK",
          "weekIndex": 0,
        },
        Object {
          "dayIndex": 2,
          "leftovers": false,
          "mealIndex": 0,
          "type": "EAT",
          "weekIndex": 0,
        },
      ],
      "dayIndex": 2,
      "mealIndex": 0,
      "weekIndex": 0,
    },
    Object {
      "actions": Array [
        Object {
          "dayIndex": 2,
          "mealIndex": 1,
          "mealType": "QUICK",
          "servings": 2,
          "type": "COOK",
          "weekIndex": 0,
        },
        Object {
          "dayIndex": 2,
          "leftovers": false,
          "mealIndex": 1,
          "type": "EAT",
          "weekIndex": 0,
        },
      ],
      "dayIndex": 2,
      "mealIndex": 1,
      "weekIndex": 0,
    },
    Object {
      "actions": Array [
        Object {
          "dayIndex": 2,
          "mealIndex": 2,
          "mealType": "QUICK",
          "servings": 2,
          "type": "COOK",
          "weekIndex": 0,
        },
        Object {
          "dayIndex": 2,
          "leftovers": false,
          "mealIndex": 2,
          "type": "EAT",
          "weekIndex": 0,
        },
      ],
      "dayIndex": 2,
      "mealIndex": 2,
      "weekIndex": 0,
    },
    Object {
      "actions": Array [
        Object {
          "dayIndex": 3,
          "mealIndex": 0,
          "type": "SKIP",
          "weekIndex": 0,
        },
      ],
      "dayIndex": 3,
      "mealIndex": 0,
      "weekIndex": 0,
    },
    Object {
      "actions": Array [
        Object {
          "dayIndex": 3,
          "mealIndex": 1,
          "type": "SKIP",
          "weekIndex": 0,
        },
      ],
      "dayIndex": 3,
      "mealIndex": 1,
      "weekIndex": 0,
    },
    Object {
      "actions": Array [
        Object {
          "dayIndex": 3,
          "mealIndex": 2,
          "type": "EAT_OUT",
          "weekIndex": 0,
        },
      ],
      "dayIndex": 3,
      "mealIndex": 2,
      "weekIndex": 0,
    },
    Object {
      "actions": Array [
        Object {
          "dayIndex": 4,
          "mealIndex": 0,
          "type": "SKIP",
          "weekIndex": 0,
        },
      ],
      "dayIndex": 4,
      "mealIndex": 0,
      "weekIndex": 0,
    },
    Object {
      "actions": Array [
        Object {
          "dayIndex": 4,
          "mealIndex": 1,
          "type": "SKIP",
          "weekIndex": 0,
        },
      ],
      "dayIndex": 4,
      "mealIndex": 1,
      "weekIndex": 0,
    },
    Object {
      "actions": Array [
        Object {
          "dayIndex": 4,
          "mealIndex": 2,
          "mealType": "NORMAL",
          "servings": 2,
          "type": "COOK",
          "weekIndex": 0,
        },
        Object {
          "dayIndex": 4,
          "leftovers": false,
          "mealIndex": 2,
          "type": "EAT",
          "weekIndex": 0,
        },
      ],
      "dayIndex": 4,
      "mealIndex": 2,
      "weekIndex": 0,
    },
    Object {
      "actions": Array [
        Object {
          "dayIndex": 5,
          "mealIndex": 0,
          "mealType": "FANCY",
          "servings": 2,
          "type": "COOK",
          "weekIndex": 0,
        },
        Object {
          "dayIndex": 5,
          "leftovers": false,
          "mealIndex": 0,
          "type": "EAT",
          "weekIndex": 0,
        },
      ],
      "dayIndex": 5,
      "mealIndex": 0,
      "weekIndex": 0,
    },
    Object {
      "actions": Array [
        Object {
          "dayIndex": 5,
          "mealIndex": 1,
          "mealType": "QUICK",
          "servings": 2,
          "type": "COOK",
          "weekIndex": 0,
        },
        Object {
          "dayIndex": 5,
          "leftovers": false,
          "mealIndex": 1,
          "type": "EAT",
          "weekIndex": 0,
        },
      ],
      "dayIndex": 5,
      "mealIndex": 1,
      "weekIndex": 0,
    },
    Object {
      "actions": Array [
        Object {
          "dayIndex": 5,
          "mealIndex": 2,
          "mealType": "FANCY",
          "servings": 2,
          "type": "COOK",
          "weekIndex": 0,
        },
        Object {
          "dayIndex": 5,
          "leftovers": false,
          "mealIndex": 2,
          "type": "EAT",
          "weekIndex": 0,
        },
      ],
      "dayIndex": 5,
      "mealIndex": 2,
      "weekIndex": 0,
    },
    Object {
      "actions": Array [
        Object {
          "dayIndex": 6,
          "mealIndex": 0,
          "type": "SKIP",
          "weekIndex": 0,
        },
      ],
      "dayIndex": 6,
      "mealIndex": 0,
      "weekIndex": 0,
    },
    Object {
      "actions": Array [
        Object {
          "dayIndex": 6,
          "mealIndex": 1,
          "type": "SKIP",
          "weekIndex": 0,
        },
      ],
      "dayIndex": 6,
      "mealIndex": 1,
      "weekIndex": 0,
    },
    Object {
      "actions": Array [
        Object {
          "dayIndex": 6,
          "mealIndex": 2,
          "mealType": "FANCY",
          "servings": 2,
          "type": "COOK",
          "weekIndex": 0,
        },
        Object {
          "dayIndex": 6,
          "leftovers": false,
          "mealIndex": 2,
          "type": "EAT",
          "weekIndex": 0,
        },
      ],
      "dayIndex": 6,
      "mealIndex": 2,
      "weekIndex": 0,
    },
  ],
  "mealsPerDay": 3,
  "startDay": 0,
  "startMeal": 0,
  "weekIndex": 0,
}
`;
